### ISSUE TITLE
Eventsub/webhook support

### ISSUE THREAD

**Issue by Serph91P** [Zibbp/ganymede#560]
Hey, would be great if you could add eventsub support. I know it requires a valid https certificate but this would greatly improve the application since this will get near to instant notification for a Livestream to start or any change made to the stream. So you could get alot more metadata faster and more precise.

Maybe you can have a look.

https://dev.twitch.tv/docs/eventsub/
https://dev.twitch.tv/docs/eventsub/handling-webhook-events/
twitchAPI.eventsub.webhook.html

**Comment by russelg** [comment]
One project that implements this is https://github.com/MrBrax/LiveStreamDVR, and I still use this just for live notifications for my own purposes.1

**Comment by Serph91P** [comment]
> One project that implements this is https://github.com/MrBrax/LiveStreamDVR, and I still use this just for live notifications for my own purposes. Also useful as it captures game and stream title changes.

Thank you. I already use this but the dev has no time to maintain at the moment and it has this wiered ram issue if you have some streams accumulated.

**Comment by russelg** [comment]
Sorry, only pointed the repo out as reference for a project which implements using these webhooks.

**Comment by Serph91P** [comment]
> Sorry, only pointed the repo out as reference for a project which implements using these webhooks.

Still thankful and am sorry if I was rude slept kinda bad.

**Comment by Zibbp** [comment]
It should be possible to use the websockets from the eventsub API. Originally I believe this feature only supported webhooks which is why I never implemented it as that would require exposing your instance publicly.

**Comment by Serph91P** [comment]
> It should be possible to use the websockets from the eventsub API. Originally I believe this feature only supported webhooks which is why I never implemented it as that would require exposing your instance publicly.

You need webhooks for it to support multiple streamers. So yes https needed.

**Comment by Zibbp** [comment]
I don't want to implement the webhook version as that would require having your Ganymede instance publicly available. This is not something I want to require.

The _websocket_ eventsub would work, but I see an immediate issue. Each "subscription" to an event costs points. Websockets have a max limit of 10 "cost" per subscription.

> The max_total_cost is 10 across all subscriptions.
https://dev.twitch.tv/docs/eventsub/manage-subscriptions/

A stream going live notification costs 1, so a max of 10 watched channels. 

> stream.online (no auth requirements, cost 1)
https://dev.twitch.tv/docs/eventsub/manage-subscriptions/


So using eventsub would limit the number of live watched channels to 10 which is not great. Unless I'm missing something about the eventsub API?

**Comment by Serph91P** [comment]
> I don't want to implement the webhook version as that would require having your Ganymede instance publicly available. This is not something I want to require.
> 
> The _websocket_ eventsub would work, but I see an immediate issue. Each "subscription" to an event costs points. Websockets have a max limit of 10 "cost" per subscription.
> 
> > The max_total_cost is 10 across all subscriptions.
> > https://dev.twitch.tv/docs/eventsub/manage-subscriptions/
> 
> A stream going live notification costs 1, so a max of 10 watched channels.
> 
> > stream.online (no auth requirements, cost 1)
> > https://dev.twitch.tv/docs/eventsub/manage-subscriptions/
> 
> So using eventsub would limit the number of live watched channels to 10 which is not great. Unless I'm missing something about the eventsub API?

yeah but cant the user decide itself? Make it an option not an requirement. You could implement all 3, and make the user choose what he wants. Like api you can have unlimited but only poll every set time. Websocket instant notification but only 10 streams. Webhook unlimited and instant but needs to be public.

Oh and websocket gets worse in my opinion: https://dev.twitch.tv/docs/eventsub/manage-subscriptions/#subscription-limits
WebSocket limits

The following limits apply per user token (client ID and user ID tuple).

    You may create a maximum of 3 WebSockets connections with enabled subscriptions. Reconnecting using a reconnection URL (see [Reconnect message](https://dev.twitch.tv/docs/eventsub/handling-websocket-events#reconnect-message)) doesn’t add to your WebSocket count.
    Each WebSocket connection may create a maximum of 300 enabled subscriptions (disabled subscriptions don’t count against the limit).
    The max_total_cost is 10 across all subscriptions.

**Comment by BigFlubba** [comment]
This is something that should be added. Like @Serph91P said, you can let the user have a choice. I currently use LiveStreamDVR, but my biggest issues are a lack of documentation (even though I figured out 75% of it) and the slow updates. You can look at my open issues to see for yourself. While it does suck having something on the open internet I have had my instance on the web behind Cloudflare using other tunnel for months without issues. Mr. Brax even said that you don't need the entire thing on the web, but as long as you have the API so it can communicate with Twitch, then it will work. While I can have the time reduced to around 30 seconds currently, I am looking for something faster with real-time updates.

Edit: Also, with Webhooks, it allows the option for you to add a graph of live viewers watching, which I really enjoy and use to see if there are any eventful or boring moments, and to see if the stream is worth it to watch. It also allows the use of logging what games the streamer is playing, like the way LiveStreamDVR does.

**Comment by Zibbp** [comment]
I do not have any plans to implement webhooks/eventsub support right now. Polling the API works plenty fine for this application. If other platforms are added they would also use the API poll method, as that is simple and always supported.

Even with polling the API, Ganymede has support for logging what games (chapters) through watched channels. Additionally the chat histogram can highlight eventful moments.

**Comment by goproslowyo** [comment]
@BigFlubba are you a developer? Would love to see a PR from you on this if you _really_ want it. Outside of contributing yourself I don't believe the current time-based live polling system is bad at all. I have mine set to THREE whole minutes and I never, never, ever have cared that the ~3 minutes were missing from the "STARTING SOON" message the streamer puts up. But again, like I said, if three minutes of missing content is that important to you I welcome your PR.

**Comment by BigFlubba** [comment]
> I do not have any plans to implement webhooks/eventsub support right now. Polling the API works plenty fine for this application. If other platforms are added they would also use the API poll method, as that is simple and always supported.
> 
> Even with polling the API, Ganymede has support for logging what games (chapters) through watched channels. Additionally the chat histogram can highlight eventful moments.

@Zibbp After a bit of letting it run, it does work pretty well. While it is not instant (or as close as you can get with Twitch) it is good for what it does. 

Is it possible that you can have it highlight the user's favorite games and such?



> @BigFlubba are you a developer? Would love to see a PR from you on this if you _really_ want it. Outside of contributing yourself I don't believe the current time-based live polling system is bad at all. I have mine set to THREE whole minutes and I never, never, ever have cared that the ~3 minutes were missing from the "STARTING SOON" message the streamer puts up. But again, like I said, if three minutes of missing content is that important to you I welcome your PR.

@goproslowyo Unfortunately, I am very bad at coding. While I have made PR's before they were simple edits that took me a long time to do. If I knew how to code better, I would love to contribute and improve this project. While I agree it's not terrible, it still has its flaws. Yes, I do usually skip the beginning of streams, not all of the ones that I record have a starting screen. Even the ones that do sometimes interact with chat while it is playing. I personally use this to not only preserve media, but with some streamers I get paid to clip and edit their content. Yes, I could go back into the VOD and grab that first 30 seconds, but sometimes the VOD gets muted because of copyright.
